lib/domain/usecase/implementation/get_movies_usecase.dart

```diff
@@ -0,0 +1,53 @@
+import 'dart:convert';
+
+import '../../../core/util/enums/endpoint.dart';
+import '../../../data/model/data_state.dart';
+import 'package:connectivity_plus/connectivity_plus.dart';
+
+import '../../entity/movie.dart';
+import '../../repository/movie_database_repository_interface.dart';
+import '../../repository/movie_repository_interface.dart';
+import '../usecase_interface.dart';
+
+class GetMoviesUseCase implements UseCase<DataState> {
+  final IMovieRepository movieRepository;
+  final IMovieDatabaseRepository movieDataBaseRepository;
+  static const String errorDataBase = 'data base failed';
+
+  GetMoviesUseCase({
+    required this.movieRepository,
+    required this.movieDataBaseRepository,
+  });
+
+  @override
+  Future<DataState<List<Movie>>> call({params}) async {
+    final connectivityResult = await (Connectivity().checkConnectivity());
+    Endpoint endpoint = params;
+    if (connectivityResult == ConnectivityResult.none) {
+      try {
+        List<Movie> movies = await movieDataBaseRepository.getMovies(
+            category: endpoint.category);
+        return DataSuccess(movies);
+      } on Exception catch (e) {
+        return DataFailed(e.toString());
+      }
+    } else {
+      DataState<List<Movie>> movies =
+          await movieRepository.getMovies(endpoint: endpoint);
+      for (var movie in movies.data!) {
+        Movie? movieInDb =
+            await movieDataBaseRepository.existsById(id: movie.id);
+        if (movieInDb != null) {
+          if (!movieInDb.categories.contains(endpoint.category)) {
+            movieInDb.categories.add(endpoint.category);
+            movieDataBaseRepository.addMovie(movieInDb);
+          }
+        } else {
+          movie.categories.add(endpoint.category);
+          movieDataBaseRepository.addMovie(movie);
+        }
```

Comment:
Maybe you can move the logic to validate if the movie exist or not on movie_database_repository

Reply:
done